test/inverse_deconv.cpp

```diff
@@ -0,0 +1,135 @@
+{
+    typedef typename cond_type<is_same_type<T, double>::value, double, float>::type OutType;
+
+    if (noDoubleTests<T>()) return;
+    if (noImageIOTests()) return;
+
+    using af::dim4;
+
+    vector<dim4>       inDims;
+    vector<string>    inFiles;
+    vector<dim_t> outSizes;
+    vector<string>   outFiles;
+
+    readImageTests(pTestFile, inDims, inFiles, outSizes, outFiles);
+
+    size_t testCount = inDims.size();
+
+    for (size_t testId=0; testId<testCount; ++testId)
+    {
+        inFiles[testId].insert(0, string(TEST_DIR "/inverse_deconv/"));
+        outFiles[testId].insert(0, string(TEST_DIR "/inverse_deconv/"));
+
+        af_array _inArray   = 0;
+        af_array inArray    = 0;
+        af_array kerArray   = 0;
+        af_array _outArray  = 0;
+        af_array cstArray   = 0;
+        af_array minArray   = 0;
+        af_array numArray   = 0;
+        af_array denArray   = 0;
+        af_array divArray   = 0;
+        af_array outArray   = 0;
+        af_array goldArray  = 0;
+        af_array _goldArray = 0;
+        dim_t nElems        = 0;
+
+        ASSERT_EQ(AF_SUCCESS, af_gaussian_kernel(&kerArray, 13, 13, 2.25, 2.25));
+
+        af_dtype otype = (af_dtype)af::dtype_traits<OutType>::af_type;
+
+        ASSERT_EQ(AF_SUCCESS, af_load_image(&_inArray, inFiles[testId].c_str(), isColor));
+        ASSERT_EQ(AF_SUCCESS, conv_image<T>(&inArray, _inArray));
+
+        ASSERT_EQ(AF_SUCCESS, af_load_image(&_goldArray, outFiles[testId].c_str(), isColor));
+        ASSERT_EQ(AF_SUCCESS, conv_image<OutType>(&goldArray, _goldArray));
+        ASSERT_EQ(AF_SUCCESS, af_get_elements(&nElems, goldArray));
+
+        unsigned ndims;
+        dim_t dims[4];
+        ASSERT_EQ(AF_SUCCESS, af_get_numdims(&ndims, goldArray));
+        ASSERT_EQ(AF_SUCCESS, af_get_dims(dims, dims+1, dims+2, dims+3, goldArray));
+
+        ASSERT_EQ(AF_SUCCESS, af_inverse_deconv(&_outArray, inArray, kerArray, gamma, algo));
+
+        double maxima, minima, imag;
+        ASSERT_EQ(AF_SUCCESS, af_min_all(&minima, &imag, _outArray));
+        ASSERT_EQ(AF_SUCCESS, af_max_all(&maxima, &imag, _outArray));
+        ASSERT_EQ(AF_SUCCESS, af_constant(&cstArray, 255.0, ndims, dims, otype));
+        ASSERT_EQ(AF_SUCCESS, af_constant(&denArray, (maxima-minima), ndims, dims, otype));
+        ASSERT_EQ(AF_SUCCESS, af_constant(&minArray, minima, ndims, dims, otype));
+        ASSERT_EQ(AF_SUCCESS, af_sub(&numArray, _outArray, minArray, false));
+        ASSERT_EQ(AF_SUCCESS, af_div(&divArray, numArray, denArray, false));
+        ASSERT_EQ(AF_SUCCESS, af_mul(&outArray, divArray, cstArray, false));
+
+        std::vector<OutType> outData(nElems);
+        ASSERT_EQ(AF_SUCCESS, af_get_data_ptr((void*)outData.data(), outArray));
+
+        std::vector<OutType> goldData(nElems);
+        ASSERT_EQ(AF_SUCCESS, af_get_data_ptr((void*)goldData.data(), goldArray));
+
+        ASSERT_EQ(true, compareArraysRMSD(nElems, goldData.data(), outData.data(), 0.03));
```

Comment:
How did you get the 0.03 threshold for this operation? Wouldn't this be depended on the type of the input and output data and its range?

Reply:
I guess what I am asking is that have you done any comparison on the image itself and looked at the results at the pixel level and calculated why there are differences. Sometimes there are subtle bugs which cannot be viewed or appear when you look at metrics like RMSD

Reply:
It would depend on the input range unless we bring them into the same range, which I am doing both while generating the test data from ITK and after calculating the output in ArrayFire tests.